src/com/lilithsthrone/game/character/GameCharacter.java

```diff
@@ -2887,8 +3007,421 @@ public boolean isSlave() {
 		return !getOwnerId().isEmpty();
 	}
 	
+	/**
+	 * Use for XML loading
+	 */
+	public void setNextPlayerSlaveryUpdateTimes(long eventUpdate, long ruleUpdate)
+	{
+		this.timeOfNextEventCheck = eventUpdate;
+		this.timeOfNextRuleCheck = ruleUpdate;
+	}
+	
+	/**
+	 * Use for exceptionally actions that are likely to cause an event to trigger, like resting. Generally just try to pad out any possible time consuming event with those prior to calling it.
+	 * 
+	 * It will delay the event's trigger for five minutes after this one or not delay it at all if it can fit in.
+	 * @param amount
+	 */
+	public void delayEventCheck(long amount)
+	{
+		long projectedDelayTime = Main.game.getMinutesPassed() + amount + GameCharacter.MINIMUM_POST_DELAY_DURATION;
+		if(projectedDelayTime > this.timeOfNextEventCheck)
+		{
+			this.timeOfNextEventCheck = projectedDelayTime;
+		}
+	}
+	
+	/**
+	 * Returns true if player has all rules at least at 0 obedience gain.
+	 */
+	public boolean isPlayerObeyingRules()
+	{
+		for(PlayerSlaveryRule slaveryRule : this.slaveOwnerRules) {
+			if(slaveryRule.getPerformanceQuality() < 0)
+			{
+				return false;
+			}
+		}
+		return true;
+	}
+	
+	/**
+	 * Handles player slavery updates
+	 */
+	public void handlePlayerSlavery(long minutesPassed)
+	{		
+		
+		boolean forcePunishment = false; // If true, forces the player to take punishment. Use for high obedience drops.
+		
+		// Checking rules compliance here
+		if(this.timeOfNextRuleCheck <= minutesPassed)
+		{
+			int totalRules = 0;
+			float obedienceChangePositive = 0;
+			float obedienceChangeNegative = 0;
+			float obedienceChangeTotal = 0;
+			
+			for(PlayerSlaveryRule slaveryRule : this.slaveOwnerRules) {
```

Comment:
```this.slaveOwnerRules.stream().mapToDouble(PlayerSlaveryRule::getPerformanceQuality).reduce(0, (sum, obedienceChange) -> {
				if ((sum < 0) != (obedienceChange < 0)){
					return Math.min(obedienceChange, sum);
				}
				return sum + obedienceChange; 
			});```

If the sum is negative, but you're adding a 0 or higher value, you want the current sum.
If the sum is negative, and you're adding a negative value, you want them added together.
If the sum is 0 or higher, and you're adding a negative value, you want the added negative value.
If the sum is 0 or higher, and you're adding a 0 or higher value, you want them added together.
sum negative != negative value thus checks for the cases "adding negative to 0 or higher" and "adding 0 or higher to negative". In those cases you want the lowest of the two (because that's the negative).
In the other cases, you want them added together.